### PR TITLE
Streaming Thrift proposal

### PR DESCRIPTION
**Pull request by prashantv**

### CODE REVIEW

File: docs/thrift.md [hunk 1]
```diff
@@ -208,3 +208,47 @@ message with error code (`code:1`) `0x05` (unexpected error).
...
+
+Streaming Thrift allows requests or responses to be streamed over TChannel.
+
+Streaming Thrift will uses a separate `as` header: `sthrift`, and is based
+on the standard Thrift protocol, with the only exception being the encoding
+of arg3 in both the `call req` and `call res`.
+
+The encoding of data for streaming arg3 is a 4 byte length-prefixed chunk:
+```
+chunk~4 chunk~4 chunk~4 ...
```
Comment (junchaowu) [inline docs/thrift.md]: How does the receiver translates the chunks to each params for arg3 If there are multi arguments for a thrift call?

File: docs/thrift.md [hunk 2]
```diff
@@ -208,3 +208,47 @@ message with error code (`code:1`) `0x05` (unexpected error).
   [`arg1`]: #arg1
   [`arg3`]: #arg3
 
+
+Streaming Thrift
```
Comment (Raynos) [inline docs/thrift.md]: Consider documenting in a seperate file. One file per "as" is nice.

File: docs/thrift.md [hunk 3]
```diff
@@ -208,3 +208,47 @@ message with error code (`code:1`) `0x05` (unexpected error).
...
+Streaming Thrift will uses a separate `as` header: `sthrift`, and is based
+on the standard Thrift protocol, with the only exception being the encoding
+of arg3 in both the `call req` and `call res`.
+
+The encoding of data for streaming arg3 is a 4 byte length-prefixed chunk:
+```
+chunk~4 chunk~4 chunk~4 ...
+```
+
+Each chunk is a Thrift encoded payload using TBinaryProtocol.
```
Comment (Raynos) [inline docs/thrift.md]: Each chunk is a Thrift encoded Struct using TBinaryProtocol

File: docs/thrift.md (hunk 1 above)
Comment (Raynos) [inline docs/thrift.md]: each chunk is a struct. so if you have a `method(1: string foo, 2: string bar, 3: string baz)`

Then you should expect each chunk to be `{ foo: ..., bar: ..., baz: ... }`

Comment (Raynos) [inline docs/thrift.md]: It notes that there is only one arg lower down.

File: docs/thrift.md [hunk 4]
```diff
@@ -208,3 +208,47 @@ message with error code (`code:1`) `0x05` (unexpected error).
...
+There is no limit to the number of chunks that can be streamed.
+
+Streaming Thrift methods have the following restrictions:
+ * If the request is streamed, the method IDL should only specify a single
```
Comment (Raynos) [inline docs/thrift.md]: Consider supporting multiple args in the form of `method(1: string bar, 2: string baz)` by saying that the request must be a struct for all the arguments.

i.e. `{ bar: '', baz: '' } { bar: '', baz: '' } ...`

File: docs/thrift.md (hunk 3 above)
Comment (Raynos) [inline docs/thrift.md]: Never mind. you allowed arbitrary types because you only support the first arg.

File: docs/thrift.md (hunk 4 above)
Comment (prashantv) [inline docs/thrift.md]: In compiled languages, we would need to generate a struct for the combination of args. While it's trivial, if the user will be required to use a struct anyway, I'd rather they create a struct explicitly.

File: docs/thrift.md (hunk 3 above)
Comment (prashantv) [inline docs/thrift.md]: Decided to only allow structs for now. Clearer how it works that way.

File: docs/thrift.md (hunk 2 above)
Comment (prashantv) [inline docs/thrift.md]: Moving to sthrift.md

**Comment by Raynos** [comment]
lgtm.

**Comment by blampe** [comment]
:shipit:

File: docs/sthrift.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+```
+chunk~4 chunk~4 chunk~4 ...
+```
+
+Arguments
+---------
+When streaming request arguments, each `arg3` chunk is the encoded payload for
+the method's first (and only) argument.
+For example, if a method defines a single argument that is a string, then
+each chunk is just a Thrift encoded string.
```
Comment (kriskowal) [inline docs/sthrift.md]: Recommend alternately just supporting streaming argument struct. Thrift methods are already accept only one argument (a struct) so going out of the way to pare out a single argument seems weird.

File: docs/sthrift.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+
+Responses
+---------
+When streaming successful (`OK`) responses, each `arg3` chunk is the encoded
+payload for the method's result type.
+For example, if a method returns a string, then each chunk should be a
+Thrift encoded string.
```
Comment (kriskowal) [inline docs/sthrift.md]: Also recommend just streaming the result struct, so {0 success} and all error cases are permitted.

Comment (Raynos) [inline docs/sthrift.md]: Streaming multiple errors is going to be confusing.

File: docs/sthrift.md (hunk 5 above)
Comment (Raynos) [inline docs/sthrift.md]: @prashantv mentioned that allowing only a single argument makes apache thrift integration simpler.

File: docs/sthrift.md (hunk 6 above)
Comment (kriskowal) [inline docs/sthrift.md]: Buying it. Paring a single entry from result and arg struct is okay by me.

**Pull request merged**